[4.0] Menus Manager: Fix system links menu items Status icon and adding badge when component is disabled

infograf768 opened the pull request:
Pull Request for Issue #24721
Replaces https://github.com/joomla/joomla-cms/pull/25185

### Summary of Changes
Correcting System Links menu items icon
Adding a badge to inform user when a component is disabled.


### Testing Instructions
Make sure you have created a System Links menu item (URL, Heading, Separator, Alias).
Create a Newsfeeds menu item and disable the Newsfeeds component in Extensions->Manage
Parameter a menu item to be published in the future.
Parameter a menu item to stop publishing in the past


### Before patch
Clicking on the status icon of a system links menu item has no effect
No way to know if the component of an existing menu item is disabled or not.


### After patch
Contrary to https://github.com/joomla/joomla-cms/pull/25185 publish_down and publish-up are respected.
If a component is disabled, the Manager will now display a badge "Component disabled"


<img width="590" alt="Screen Shot 2019-06-13 at 13 53 17" src="https://user-images.githubusercontent.com/869724/59431620-9b0f2900-8de5-11e9-893f-5852811a39f9.png">


### Note
Status icon tip as well as title tip display are not modified as this requires love and care in another PR.

@810 @SharkyKZ

infograf768 committed:
[4.0] Menus Manager: Fix system links menu items Status icon and adding
badge when component is disabled

810 commented:
I have tested this item :white_check_mark: successfully on c0c48f5512d7e95fbb05b75c0fef9a16d46c7bb0<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/25197">issues.joomla.org/tracker/joomla-cms/25197</a>.</sub>

infograf768 committed:
cs

alikon commented on administrator/components/com_menus/Service/HTML/Menus.php:
```diff
@@ -260,4 +121,32 @@ public function visibility($params)
+
+			$query->select($db->quoteName('a.enabled'))
+				->from($db->quoteName('#__extensions', 'a'))
+				->where($db->quoteName('a.element') . ' = ' . $db->quote($componentname));
```
```suggestion
				->where($db->quoteName('a.element') . ' = :element')
				    ->bind(':element', $componentname);
```

all queries should be prepared statement

SharkyKZ commented:
This is very inefficient as we're running a DB query for every item.

infograf768 commented:
Tried my best. I guess I will just give up. No idea how to make it more efficient.
As for “prepared statement” use, would it make the query more efficient ?

SharkyKZ commented:
If we're fine with removing custom strings/states here, select component status (`e.enabled`) in the model query and use that.

Otherwise, the complete solution is probably to copy-paste in `JHtmlJgrid::published()` into `Joomla\Component\Menus\Administrator\Service\HTML::published()` and adapt it to custom states.

alikon commented:
> As for “prepared statement” use, would it make the query more efficient ?

unfortunately not,  it help to prevent SQL Injection

infograf768 committed:
Using already existing query instead of creating a new method

infograf768 commented:
@SharkyKZ 
Corrected by using existing model query. Adding the publish_down and publish_up to the custom strings/state would just make such a tip too complicated.

@alikon
As I deleted the new method, no use anymore for a prepare statement.

@810 
Sorry, please test again.

SharkyKZ commented on administrator/components/com_menus/tmpl/items/default.php:
```diff
@@ -206,6 +206,11 @@
 											  title="<?php echo isset($item->item_type_desc) ? htmlspecialchars($this->escape($item->item_type_desc), ENT_COMPAT, 'UTF-8') : ''; ?>">
 											<?php echo $this->escape($item->item_type); ?></span>
 									</div>
+									<?php if (!empty($item->componentname) && $item->enabled === 0) : ?>
```
`$item->enabled` can be a string or an integer. And `NULL` for non-existing components.

infograf768 replied:
Suggestion ?

infograf768 replied:
`$item->enabled != 1`   ?

SharkyKZ replied:
I propose this to differentiate between disabled and non-existing components:
```
<?php if ($item->type === 'component' && !$item->enabled) : ?>
	<span class="badge badge-secondary"><?php echo Text::_($item->enabled === null ? 'COM_MENUS_LABEL_COMPONENT_NOT_FOUND' : 'COM_MENUS_LABEL_COMPONENT_DISABLED'); ?></span>
<?php endif; ?>
```
And move it after line 202 so it appears next to `Hidden` label, if such exists.

infograf768 replied:
We have discussed elsewhere the placement of the Note, and consequently the placement of the Hidden Label. Both need care as they are displayed on the same line as the Alias which makes it real confusing.
It is on purpose therefore that I added the new badge in a separate `<div>` displayed last. Hopefully, a clever admin, at seeing the new badge, will correct it by reinstalling the component or deleting the useless menu item, therefore such a badge should be temporary while Hidden is always of use.
We have an existing string to replace the new `COM_MENUS_LABEL_COMPONENT_NOT_FOUND` proposed.
It is
`JLIB_APPLICATION_ERROR_COMPONENT_NOT_FOUND="Component not found."`
Will correct PR to use your code.

infograf768 replied:
Thanks for the code.
please test

infograf768 committed:
taking unexisting component into account

810 commented:
I have tested this item :white_check_mark: successfully on 409e60fc8a73272148b3d100235d96c4e6a2396a<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/25197">issues.joomla.org/tracker/joomla-cms/25197</a>.</sub>

Quy commented:
The following were removed but not in the language file.

> COM_MENUS_EXTENSION_UNPUBLISHED_ENABLED
> COM_MENUS_EXTENSION_PUBLISHED_DISABLED
> COM_MENUS_EXTENSION_UNPUBLISHED_DISABLED

Quy commented:
Please advise if issue #19874 is related.

infograf768 committed:
Deleting 4 more unused strings. Tks @Quy

infograf768 committed:
Merge branch '4.0-dev' into 4.0_menumanager_icon_disabled

infograf768 commented:
> Please advise if issue #19874 is related.

I do not think so. Here is the process:

In 3.x we use the override `menus.state` (and publish_down as well as publish_up do not exist).

In case the component exists, is disabled and the item is Published we have a `warning` icon.

<img width="473" alt="Screen Shot 2019-06-15 at 08 31 17" src="https://user-images.githubusercontent.com/869724/59547973-38826e00-8f48-11e9-949d-782d7eac3f6b.png">

In case the component exists, is disabled and the item is Unpublished we have a `trash` icon.

<img width="595" alt="Screen Shot 2019-06-15 at 08 36 38" src="https://user-images.githubusercontent.com/869724/59548000-bcd4f100-8f48-11e9-9a7d-b0fb0cc501e6.png">

In 4.0, with this PR, we get an Unpublished icon in both cases WITH the new badge when the component is disabled and specific icons for published/unpublished or publish_up and publish_down when enabled.

Filtering by Published/Unpublished will work fine whatever the status of the component.

The case of the unexisting component ( #19874 ), **this both for 3.x and 4.0**, is not dealt with the same way: it is done in the View itself. It defaults to using the Unpublished icon and adds the following text below the Menu Item title as it just sets the value to `$item->item_type`:

```
else
					{
						if (preg_match("/^index.php\?option=([a-zA-Z\-0-9_]*)/", $item->link, $result))
						{
							$value = JText::sprintf('COM_MENUS_TYPE_UNEXISTING', $result[1]);
						}
						else
						{
							$value = JText::_('COM_MENUS_TYPE_UNKNOWN');
						}
					}
					break;
```
This PR adds a badge "Component not found." in this case for 4.0.

Basically `JHtmlJGrid` does not deal either with enabled or unexisting.

Quy commented:
Should the badge display at the same location as `Hidden`?

infograf768 commented:
@Quy 
> Should the badge display at the same location as Hidden?

I have replied on this here:
https://github.com/joomla/joomla-cms/pull/25197#discussion_r293681797

infograf768 committed:
Merge branch '4.0-dev' into 4.0_menumanager_icon_disabled

Quy commented:
I have tested this item :white_check_mark: successfully on 29318a114dcca34608971f693a93a91aa2b32409<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/25197">issues.joomla.org/tracker/joomla-cms/25197</a>.</sub>

infograf768 commented:
RTC as last changes since @810  tested concerned only deleting some unsused lang strings.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/25197">issues.joomla.org/tracker/joomla-cms/25197</a>.</sub>

infograf768 committed:
Merge branch '4.0-dev' into 4.0_menumanager_icon_disabled

infograf768 committed:
Merge branch '4.0-dev' into 4.0_menumanager_icon_disabled

infograf768 committed:
Merge branch '4.0-dev' into 4.0_menumanager_icon_disabled

infograf768 committed:
Merge branch '4.0-dev' into 4.0_menumanager_icon_disabled

infograf768 commented:
@wilsonge 
Merge?

roland-d commented:
Thank you.
